spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java

```diff
@@ -308,6 +339,17 @@ public Scan buildChangelogScan() {
     return new SparkChangelogScan(spark, table, scan, readConf, expectedSchema, filterExpressions);
   }
 
+  private Long getStartSnapshotId(Long startTimestamp) {
+    Snapshot oldestSnapshotAfter = SnapshotUtil.oldestAncestorAfter(table, startTimestamp);
+    Preconditions.checkArgument(
```

Comment:
We don't have to throw exception here, another option is to return an empty set of row.

Reply:
To me, may be slightly more user friendly to return empty set.  But open.

Reply:
My question would be do we need to differentiate between a snapshot existed yet had no changes and no such snapshot existed.I think i'm good with the exception because it makes it clear you aren't getting results because of the input parameters and not because of the table.

Reply:
> do we need to differentiate between a snapshot existed yet had no changes and no such snapshot existed.

I will consider these two use cases.
1. Query with snapshot id. We should differentiate it, throwing an exception makes sense here.
2. Query only with time range. We may assume user doesn't have to know the concept of `snapshot`. We may not differentiate it, returning an empty set makes more sense.

Subtle difference though. I am fine with either option.

Reply:
Having an exception to distinguish is a fair point.  Initially I was thinking it would be a pain for user to have to deal with exception if they wanted to do all changes from a random time, but they can find the earliest snapshot and make sure the timestamp is appropriate to avoid exception.  So still no real preference, ok either way.

Reply:
I think you are correct @flyrain , I only really would want the error for 1.